### PR TITLE
Fix errors limits

### PR DESCRIPTION
**Pull request by edwintorok**
Fix for #107, #109, #112 and updated tests meant for discussion.

This is a fairly large change (especially the unification), I tried to do it in incremental steps, but please tell me how I should proceed, or if you'd like a different approach.
The reason for the unification is that I kept fixing the same bugs over 3 implementations (unix,  unix+ssl, unix+tls) and one bug got fixed just by unifying the implementations (#114).

Depends on https://github.com/mirage/ocaml-conduit/pull/115 and https://github.com/mirage/ocaml-conduit/pull/111, so please don't merge yet. Once those 2 other PRs are decided I'll rebase this one.

### CODE REVIEW

File: lib/conduit_lwt_unix_ssl.ml [hunk 1]
```diff
@@ -16,34 +16,23 @@
...
-  Lwt.catch (fun () -> f fd) (fun e ->
-      Lwt.catch (fun () -> Lwt_unix.close fd) (fun _ -> return_unit) >>= fun () ->
-      fail e
-    )
-
 module Client = struct
   (* SSL TCP connection *)
-  let t = Ssl.create_context Ssl.TLSv1 Ssl.Client_context
+  let default_ctx = Ssl.create_context Ssl.SSLv23 Ssl.Client_context
+  let () = Ssl.disable_protocols default_ctx [Ssl.SSLv23]
```
Comment (avsm) [inline lib/conduit_lwt_unix_ssl.ml]: why this change from `Ssl.TLSv1` to `Ssl.SSLv23` followed by the disabling of the older protocol?

**Comment by avsm** [comment]
This is absolutely the right direction, thanks very much! Will review shortly...

File: lib/conduit_lwt_unix_ssl.ml (hunk 1 above)
Comment (edwintorok) [inline lib/conduit_lwt_unix_ssl.ml]: Before OpenSSL 1.1.0 the only way to enable TLS1+ is to use `SSLv23_method` which enables all protocols,
and then selectively disable the protocols you don't want ... ugly, see https://github.com/mirage/ocaml-conduit/issues/102 and https://www.openssl.org/docs/manmaster/ssl/SSL_CTX_new.html.
After OpenSSL 1.1.0 there is a nicer API https://www.openssl.org/docs/manmaster/ssl/SSL_CTX_set_min_proto_version.html that sets protocol version limits but `ocaml-ssl` doesn't support that yet.

I didn't want to enable just `TLSv1.2` because when TLSv1.3 appears it wouldn't be automatically supported.

**Comment by vbmithr** [comment]
@avsm @rgrinberg 
What's the status of this? I think we should go ahead, review and hopefully merge.
Some bugs have been there for too long…

**Comment by vbmithr** [comment]
Ok, it depends on #111.

**Comment by vbmithr** [comment]
@edwintorok Hi Edwin, sorry for the lag. Could you go ahead on this?

**Comment by edwintorok** [comment]
I'll need to rebase/redo some of the commits, the underlying code has changed quite a lot (in fact some unification already happened https://github.com/mirage/ocaml-conduit/commit/e555367a0a887eda598f97294222edfad5188a0c which is good). I'll update this PR when I'm done.

**Comment by vbmithr** [comment]
On 06/12/2016 11:10, Török Edwin wrote:
> I'll update this PR when I'm done.

Great! Thank you very much.

**Comment by rgrinberg** [comment]
I've just completed most of the unification work in #173. So let's just get the bug fixes from this PR + the good interface changes.

**Comment by edwintorok** [comment]
Thanks @rgrinberg, your unification reduced the number of commits in this PR significantly, I pushed a simplified branch. I have updated the tests (in `tests/unix/cdtest`) and they pass, however I'd like to do some more tests to make sure that all the bugs referenced in this PR are indeed gone.

I'd  also like to run `tests/unix/run.sh` and `tests/unix/run_tls.sh` with Travis, do you have some pointers on what I need to modify in `.travis.yml` to make that happen?

**Comment by rgrinberg** [comment]
Thanks for working on this. Let me know how I can help.

>I'd also like to run `tests/unix/run.sh` and `tests/unix/run_tls.sh` with
>Travis, do you have some pointers on what I need to modify in `.travis.yml` to
>make that happen?

I think you should just create a make target that exercises all the tests and
then add this command to `build-test` in the opam file.

That should build the tests automatically.

**Comment by rgrinberg** [comment]
Hmm, how come lwt.ssl isn't being install on travis?

# ocamlbuild -use-ocamlfind -classic-display -cflags "-w A-4-33-40-41-42-43-34-44" cdtest.native
# ocamlfind ocamldep -package lwt.ssl -package lwt.unix -package tls.lwt -package ipaddr.unix -package sexplib -package ipaddr -package cstruct -package uri -package stringext -package logs -modules tests/unix/cdtest.ml > tests/unix/cdtest.ml.depends
# + ocamlfind ocamldep -package lwt.ssl -package lwt.unix -package tls.lwt -package ipaddr.unix -package sexplib -package ipaddr -package cstruct -package uri -package stringext -package logs -modules tests/unix/cdtest.ml > tests/unix/cdtest.ml.depends
# ocamlfind: Package `lwt.ssl' not found
# Command exited with code 2.
# Makefile:74: recipe for target 'cdtest' failed
### stderr ###
# make: *** [cdtest] Error 10

**Comment by avsm** [comment]
Is it 4.04? No ssl or tls there yet last I looked

**Comment by rgrinberg** [comment]
4.03 is also erroring out like this (and the other builds too)

**Comment by edwintorok** [comment]
Is there a phase in the travis run where the depopts aren't installed yet? (I don't see anything in the build logs about building ssl or async). Anyway I should add some conditionals on that test, will try to run that travis  test script locally to see what happens.

**Comment by edwintorok** [comment]
Added the conditionals, the tests pass, but they are not actually run, looks like the tests are not re-run after installing the depopts.

Also found a problem with the `cdtest_tls` test: it is not deterministic under high CPU load (i.e. `( while true; do  echo -n;done )` on all your cores) it will fail with an error when the client tries to read the message sent by the server:

```
cdtest_tls.native: main: Server running on port 9192
cdtest_tls.native: main: Server: Callback started.
cdtest_tls.native: main: Connected!
cdtest_tls.native: main: Written hello.
cdtest_tls.native: main: Server: read hello.
cdtest_tls.native: main: Got correct msg, disconnecting.
cdtest_tls.native: main: Server: Callback started.
cdtest_tls.native: main: Connected!
cdtest_tls.native: main: Written hello.
cdtest_tls.native: main: Server: read hello.
./cdtest_tls.native: "Tls_lwt.write" failed: Broken pipe
```

Doing a graceful shutdown on the server with `Tls_lwt.of_t ~close` works this around, but doesn't seem right:
```ocaml
  Lwt.catch (fun () -> Tls_lwt.Unix.read t (Cstruct.create 1) >|= ignore)
     (fun _ -> Lwt.return_unit) >>= fun () ->
  Tls_lwt.Unix.close t
```

**Comment by rgrinberg** [comment]
Perhaps we can list the depopts as `{test}` dependencies as well?

**Comment by edwintorok** [comment]
The tests run now, but some fail: debian-stable due to the `Tls_lwt` shutdown problem, centos-6 due to ctypes segfault (reported), fedora-24/alpine due to lack of m4 (probably because I pinned ocaml-ci-scripts to HEAD to run this test).

If you prefer I can remove the `.travis.yml` changes from this PR, and put them into a separate PR and enable it only once all tests pass.

File: lib/conduit_lwt_server.mli [hunk 2]
```diff
@@ -1,6 +1,8 @@
 
 val close : 'a Lwt_io.channel * 'b Lwt_io.channel -> unit Lwt.t
 
+val set_max_active : int -> unit
```
Comment (rgrinberg) [inline lib/conduit_lwt_server.mli]: Is there any way to do this on a per server basis? This doesn't seem like a constant that can be universal for all servers started with conduit.

**Comment by rgrinberg** [comment]
I've cherry picked the first commit and your lwt_tls fix will also be merged shortly. So feel free to rebase after that.

Would be nice to have the CI changes in a separate PR but we're likely not to wait and just merge them straight away. We must read the travis often enough anyway and false negatives are better than false positives.

File: lib/conduit_lwt_server.mli (hunk 2 above)
Comment (edwintorok) [inline lib/conduit_lwt_server.mli]: It would be better to leave this decision up to the user of conduit, you might want to have a global limit, per server limit or some combination of them. I went with a global limit here because the underlying resource (process file descriptors) are a global resource too (unless you're running Mirage I guess).
How about  an optional callback argument for `val serve` instead, which would get called with the number of active connections on the current server, the number of active connections globally, and decides what to do with the connection  (accept it, or sleep until either of these two changes).

Comment (rgrinberg) [inline lib/conduit_lwt_server.mli]: I see, that makes sense. But although it's a global resource we can only share this limit between conduit servers here.  

Anyway, I say let's leave it as is for now. If we need more flexibility, then we can consider your proposal here. Seems like adding this limit was motivated by a concrete need you had so let's wait until something concrete like that pops up.

**Comment by rgrinberg** [comment]
This PR is ready after rebase.

**Comment by edwintorok** [comment]
Thanks, pushed a rebased branch, lets wait until travis finishes.

**Comment by edwintorok** [comment]
Ok, its got the m4 failure from last time (not the same ones because meantime ocaml-ci-scripts got reverted back to not running the depopts tests until some other problems are fixed). If you are OK with having false negatives in the CI for now, then please go  ahead and merge it.

**Comment by rgrinberg** [comment]
Yeah, this is good enough for now.
